x-pack/plugins/security/server/session_management/session_index.ts

```diff
@@ -484,24 +489,58 @@ export class SessionIndex {
       });
     }
 
+    const operations: Array<Required<Pick<BulkOperationContainer, 'delete'>>> = [];
     try {
-      const { body: response } = await this.options.elasticsearchClient.deleteByQuery(
-        {
-          index: this.indexName,
-          refresh: true,
-          body: { query: { bool: { should: deleteQueries } } },
-        },
-        { ignore: [409, 404] }
-      );
-
-      if (response.deleted! > 0) {
-        this.options.logger.debug(
-          `Cleaned up ${response.deleted} invalid or expired session values.`
+      const { body: searchResponse } =
+        await this.options.elasticsearchClient.search<SessionIndexValue>(
+          {
+            index: this.indexName,
+            body: { query: { bool: { should: deleteQueries } } },
+            _source_includes: 'usernameHash,provider',
+          },
+          { ignore: [409, 404] }
         );
-      }
```

Comment:
Ah, good point, thanks! I mixed up the default with the maximum pagination size. 

I've added pagination now so we're removing sessions in batches.

Reply:
Hold on, I'm deleting the items that are being returned by the first batch. So next time I run the search I'm getting the next 1,000 sessions. I'm repeating that until there are no more sessions left in the index. Am I misunderstanding this?

Reply:
Oh, I skimmed your changes and misread -- you're searching _and_ deleting in each loop.

I had it in my head to page through _all_ the objects _then_ delete. Sorry about that. That's what I was implying with `refresh: false`, because if you don't refresh the index then your subsequent searches won't return fresh results.

You don't want to loop it like it's currently structured (search + delete in a single iteration) because then you'll be forced to use `refresh: 'wait_for'` each time you delete, which will really hurt performance.

Reply:
I've added in pagination. Please have another look.